packages/@aws-cdk/assertions/lib/match.ts

```diff
@@ -1,13 +1,242 @@
 import { ABSENT } from './vendored/assert';
 
 /**
- * Partial and special matching during template assertions
+ * Denotes a failure when matching patterns to targets.
  */
-export class Match {
+export interface MatchFailure {
```

Comment:
I wonder if it's worth encoding this into a class...

Reply:
I have a feeling `MatchFailures` need to encode the entire matching tree... a flat list is not going to give enough information once there are alternatives to try out.

Reply:
All the alternatives will be tried out and the 'closest' match will be recorded. That's my design at the moment.

Reply:
Yes for top-level matches between resources. But not INSIDE matchers where multiple alternatives are tried (like in `arrayWith()`)